src/main/java/sparksoniq/spark/closures/ForClauseFlatMapClosure.java

```diff
@@ -20,36 +20,34 @@
 package sparksoniq.spark.closures;
 
 import org.apache.spark.api.java.function.FlatMapFunction;
-import org.apache.spark.api.java.function.Function;
-import org.apache.spark.api.java.function.MapFunction;
 import org.apache.spark.sql.Row;
 import org.apache.spark.sql.RowFactory;
 import org.apache.spark.sql.types.StructType;
 import sparksoniq.jsoniq.item.Item;
 import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
 import sparksoniq.semantics.DynamicContext;
-import sparksoniq.spark.SparkSessionManager;
 
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 
 public class ForClauseFlatMapClosure implements FlatMapFunction<Row, Row> {
     RuntimeIterator _expression;
+    StructType _oldSchema;
 
-    public ForClauseFlatMapClosure(RuntimeIterator expression) {
+    public ForClauseFlatMapClosure(RuntimeIterator expression, StructType oldSchema) {
         this._expression = expression;
+        this._oldSchema = oldSchema;
     }
 
     @Override
     public Iterator<Row> call(Row row) {
-        StructType schema = row.schema();
```

Comment:
Could you explain what is wrong (or what is a bug or a known bug) with row.schema()? It sounds intriguing indeed that this doesn't work properly, as DataFrames documents afaik that rows carry their schema. Or maybe this has to do with the way they are initialized since we are making our own rows.

Reply:
Your guess is very much on point. When dataframes are created through createDataFrame function the rows indeed accurately carry the schema information. I believe the rows are attached the schema information during this function's call. 

In our case, DF manipulation in the consequent operations is carried out in the following form:
```
return df.flatMap(new ForClauseFlatMapClosure(_expression, oldSchema), RowEncoder.apply(newSchema));
``` 
New rows are created within the closure and the RowEncoder.apply(newSchema) is applied on the results. This applies the new schema at the dataframe level, but individual rows don't contain the schema information. Fortunately for our use case, this information is not needed at the row level as we can simply pass the schema information to the closure.

Reply:
I understand, thank you for the precision. I think this hack is fine for now, but it is worth opening an issue to investigate it fully later on. It may actually even "solve itself" when we harness Spark SQL.